### DISCUSSION TITLE
njit to speed up functions on awkward arrays

### DISCUSSION THREAD

**Discussion by fspinna** [scikit-hep/awkward#1824] [Q&A] [answered]
Hi everyone,
I'm relatively new to Numba and even newer to awkward arrays. I'm trying to understand how to use njitted functions with awkward arrays as input. For example, let's say that I've a list of numpy arrays having different lengths:

```python
import numpy as np
import numba as nb
import awkward as ak

n_rep = 100
lists = [np.array([1, 2, 3]), np.array([1,2])] * n_rep
ak_array = ak.Array(lists)
lists_numba = nb.typed.List(lists)  # numba typed list
```

I want to take the mean with axis=1. Of course, I can do it natively with awkward arrays, but to do it with numba I have to iterate over each np.array in the list, so the function will look like something like this:

```python
@nb.njit
def nb_mean(arr):
  means = np.empty(shape=len(arr))
  for i in range(len(arr)):
    means[i] = np.mean(arr[i])
  return means

nb_mean(lists_numba)
```
If I pass the awkward array to this function, numba raises this error:

```python
TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function mean at 0x7f05dd254b00>) found for signature:
 
 >>> mean(ak.ArrayView(ak.NumpyArrayType(array(int64, 1d, A), none, {}), None, ()))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'Numpy_method_redirection.generic': File: numba/core/typing/npydecl.py: Line 379.
    With argument(s): '(ak.ArrayView(ak.NumpyArrayType(array(int64, 1d, A), none, {}), None, ()))':
   Rejected as the implementation raised a specific error:
     TypeError: array does not have a field with key 'mean'
   
   (https://github.com/scikit-hep/awkward-1.0/blob/1.10.1/src/awkward/_connect/_numba/layout.py#L339)
  raised from /usr/local/lib/python3.7/dist-packages/awkward/_connect/_numba/layout.py:339

During: resolving callee type: Function(<function mean at 0x7f05dd254b00>)
During: typing of call at <ipython-input-27-1014e122c504> (8)


File "<ipython-input-27-1014e122c504>", line 8:
def nb_mean(arr):
    <source elided>
  for i in range(len(arr)):
    means[i] = np.mean(arr[i])
    ^
```

The code works if I first convert the awkward array with np.array as in the following:

```python
@nb.njit
def nb_ak_mean(arr):
  means = np.empty(shape=len(arr))
  for i in range(len(arr)):
    means[i] = np.mean(np.array(arr[i]))  # convert first to np.array then apply np.mean
  return means
```

So basically, If I try to pass directly ak_array to the `np.mean` without first converting it with `np.array` function, numba raises an error. 
These are the performance of the 3 approaches:

```python
%timeit ak.mean(ak_array, axis=1)
%timeit nb_ak_mean(ak_array)
%timeit nb_mean(lists_numba)

1.42 ms ± 131 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # native awkward
74 µs ± 1.33 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # numba + awkward
7.61 µs ± 41.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)  # numba + List
```

The approach using the List object from numba seem to be significantly faster than using awkward arrays. Am I doing something wrong?

**Comment by ianna** [comment]
@fspinna - which version of awkward do you use?
```python
>>> ak.__version__
'2.0.0rc1'
```

**Reply by fspinna** [reply]
numba == 0.56.3
awkward == 1.10.1
numpy == 1.21.6

**Comment by agoose77** [comment, marked as answer]
Awkward Array (and Numba's) performance can be measured approximately as `time = initial_cost + rate*amount_of_work`. In this case, your array is too small for the work-scaling to be properly measured, i.e. the setup costs (`initial_cost`) are dominating the performance.

If you change your array to be more like ~1,000,000 elements, then the performance difference between the two Numba jitted cases is closer to a factor of 4:[^caveat]
```
%timeit ak.mean(ak_array, axis=1)
%timeit nb_ak_mean(ak_array)
%timeit nb_mean(lists_numba)
763 ms ± 19.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
117 ms ± 9.63 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
28.8 ms ± 1.55 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The rest of this difference probably stems from the use of `np.array()` instead of `np.asarray()`. In Numba's case, we have a faster implementation of array conversion for `asarray` than `array`; `asarray` only handles 1D contiguous arrays, and directly interprets them as a numba `Array`, whereas the `np.array` function performs a copy (it can handle >1D arrays, and these can be non-contiguous in Awkward). This closes the performance gap, and now the Awkward jitted case is (barely) faster than the numba list jitted case:
```
%timeit nb_ak_mean(ak_array)
18.3 ms ± 703 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Of course, Numba isn't always able to eek out the best performance if you use NumPy operations. A bare loop is usually the fastest solution, and can help avoid these kinds of quirks.

In general, though, don't worry about this kind of performance difference if it's not bottlenecking your workflow. There are always things you can do to improve performance, e.g. handling raggedness explicitly in your kernel by flattening the array and passing in the sublist lengths, but it's not always worth the extra maintenance burden and code complexity.

[^caveat]: Note that this isn't a perfect test; Awkward Array should (in general) have a better memory layout than multiple small NumPy arrays; Awkward usually represents raggedness as variable-length views over a contiguous buffer (certainly not always, but often). This represents a more favourable scenario in terms of memory access.

**Reply by fspinna** [reply]
Thanks! Perfectly clear. This fits exactly my use case. Tested on my machine and your version is significantly faster than the numba List version.

**Reply by jpivarski** [reply]
That's right: if the total time for a Numba-JITted function to run is µs, we don't know how it will scale (the `rate` part) until we give it a larger problem. This is likely dominated by `initial_cost`.

--------------

Another consideration is that Python lists defined like

```python
lists = [np.array([1, 2, 3]), np.array([1, 2])] * n_rep
```

are actually repeated references to the same two arrays, not a dataset consisting of 5 × `n_rep` numbers (as the Awkward Array is, since the ak.Array constructor iterates over the data, copying it into a columnar representation).

These two small arrays are easily small enough to fit in a CPU's cache (even the L1 cache, which is typically between 8 kB and 64 kB), so it does not need to pull data from main memory in the calculation. If the calculation you're performing is simple enough, getting data from main memory can easily become the bottleneck.

Getting estimates from [this infographic](http://ithare.com/infographics-operation-costs-in-cpu-clock-cycles/),

  * the `+` in your `np.mean` takes < 1 clock cycle
  * the floating point `/` in your `np.mean` takes 10‒40 clock cycles (`np.sum` would be very different from `np.mean`!)
  * fetching data from L1 takes 3‒4 clock cycles
  * fetching data from main memory takes 100‒150 clock cycles

(Note: that's _after_ you already have an `n_rep` large enough that the process isn't dominated by unboxing/boxing data as you enter and leave the Numba context. I'm assuming you've already applied @agoose77's correction.)

Naively, clock ticks translate into real time by dividing by your clock speed (~3 GHz), but even this is complicated by the fact that the CPU actually runs the process in pipelines and speculatively evaluates based on guesses that can later be invalidated. The main takeaway from these numbers is the relative orders of magnitude.

Very likely, the real dataset you're interested in is not the same two Python objects repeated many times; I'll bet it's a large set of distinct lists, all with different values and lengths. While `%timeit` is correctly measuring the rate of _this_ test, you're probably interested in a different test. (Supposedly objective performance testing is full of subjective choices!)

--------

Another thing to be aware of is that this line:

```python
    means[i] = np.mean(np.asarray(arr[i]))
```

does not actually copy the data in the Awkward Array, so it's safe to use even if `arr[i]` is big. (Note that this is `np.asarray`, not `np.array`, as @agoose77 pointed out.) It has to create an [ArrayModel struct](https://github.com/numba/numba/blob/f4b6fe0124fa7734ba2bb28c7685db5b6306c486/numba/core/datamodel/models.py#L871-L884) on the stack, which is something like 64 bytes (depending on how many dimensions the array has), but that's a constant-size metadata specifying pointers to the object, data type, `shape` and `strides`, not the variable-sized data _in_ the array. It should be _relatively_ fast. (I can't put a number on it.)

It would be nice if you didn't have to do the explicit cast at all; that's open issue #509.

But as @agoose77 pointed out, vectorized functions inside of the JIT-compiled context aren't advantageous as they are in pure Python. You can sometimes do better with an explicitly defined function because LLVM will recognize that you don't need the full features of an `np.ndarray` object, just the ability to iterate over the contents, and it might be able to optimize that down into fewer instructions. I haven't tested it, but this might be a faster `nb_ak_mean`:

```python
@nb.njit
def nb_ak_mean(arr):
  means = np.zeros(shape=len(arr))
  for i, x in enumerate(arr):
    for j in range(len(x)):
      means[i] += x[j]
    means[i] /= len(x)   # maybe you'll get some NaNs from division-by-zero
  return means
```

Refactoring this into helper functions (one iterates, the other computes the mean) shouldn't be any slower because `@nb.njit` functions that call `@nb.njit` functions skip the boxing/unboxing steps and may even be inlined by LLVM. Doing it in one function vs. multiple functions is not a performance/readability tradeoff, but functions that require an ArrayModel struct like `np.mean` do have to do the extra work of making that struct, small as it may be.

**Reply by jpivarski** [reply]
> The rest of this difference probably stems from the use of `np.array()` instead of `np.asarray()`. In Numba's case, we have a faster implementation of array conversion for `asarray` than `array`; `asarray` only handles 1D contiguous arrays, and directly interprets them as a numba `Array`, whereas the `np.array` function performs a copy (it can handle >1D arrays, and these can be non-contiguous in Awkward).

I missed this and it's absolutely true: `np.array` is a much slower implementation than `np.asarray`; it's more general and does more runtime conversion. What I was saying applies to `np.asarray`. In fact, I'm going to change history and fix the text in my comment.

--------------

For reference, these are the `np.array` and `np.asarray` implementations. The `np.array` builds up a lot of runtime code to check the sizes of iterables, to see if they are regular enough to be converted into an n-dimensional NumPy array, whereas the `np.asarray` implementation takes compile-time information and only applies to regular-typed arrays. (In fact, it should someday be expanded to regular n-dimensional arrays; this `np.asarray` only works on 1D arrays.)

https://github.com/scikit-hep/awkward/blob/2bdd11426b83ff52c6a3ecad4c64d1841663e9f9/src/awkward/_connect/numba/arrayview.py#L856-L1002

This Python code runs during Numba's compilation phase.

* `overload_np_array` creates Python code in strings and evaluates them as a new Numba function to check the size of the Awkward iterable—whatever its type may be—allocates a new NumPy array, and copies the data into that new array. That's expensive!
* `type_asarray` checks the type information of the Awkward Array: it can't be a variable-length list
* `lower_asarray` generates LLVM bytecode that only builds the ~64 byte metadata for the lowered NumPy array.

**Reply by fspinna** [reply]
Thanks a lot for the details. Very interesting indeed!
